File: python/design-patterns/builder.py
import threading

class Product:
    """Represents the complex object being constructed."""
    def __init__(self):
        self.parts = []
        self.mutex = threading.Lock()

    def add_part(self, part):
        """Method to add a part to the product."""
        with self.mutex:
            self.parts.append(part)

    def show_parts(self):
        """Method to display the parts of the product."""
        with self.mutex:
            print("Product parts:", ", ".join(self.parts))

class Builder:
    """Interface for creating parts of a complex object."""
    def build_part_a(self):
        """Method to build part A."""
        raise NotImplementedError

    def build_part_b(self):
        """Method to build part B."""
        raise NotImplementedError

    def get_product(self):
        """Method to retrieve the constructed product."""
        raise NotImplementedError

class ConcreteBuilder(Builder):
    """Implements the Builder interface to construct a specific product."""
    def __init__(self):
        self.product = Product()

    def build_part_a(self):
        """Method to build part A."""
        self.product.add_part("Part A")

    def build_part_b(self):
        """Method to build part B."""
        self.product.add_part("Part B")

    def get_product(self):
        """Method to retrieve the constructed product."""
        return self.product

class Director:
    """Responsible for using the builder to construct a complex object."""
    def __init__(self, builder: Builder):
        self.builder = builder

    def construct(self):
        """Method to construct the product using the assigned builder."""
        self.builder.build_part_a()
        self.builder.build_part_b()

def main():
    """Entry point of the program."""
    builder = ConcreteBuilder()
    director = Director(builder)

    threads = []

    def thread_func():
        """Function to be executed by each thread."""
        director.construct()
        product = builder.get_product()
        product.show_parts()

    for _ in range(5):
        t = threading.Thread(target=thread_func)
        threads.append(t)
        t.start()

    for thread in threads:
        thread.join()

if __name__ == "__main__":
    main()

File: cpp/design-patterns/builder/builder.cpp
#include <iostream>
#include <memory>
#include <mutex>
#include <thread>
#include <vector>

/**
 * @brief The Product class represents the complex object being constructed.
 */
class Product {
public:
    /**
     * @brief Method to add a part to the product.
     * 
     * @param part The part to add.
     */
    void addPart(const std::string& part) {
        std::lock_guard<std::mutex> lock(mutex_);
        parts += part + ", ";
    }

    /**
     * @brief Method to display the parts of the product.
     */
    void showParts() {
        std::lock_guard<std::mutex> lock(mutex_);
        std::cout << "Product parts: " << parts << std::endl;
    }

private:
    std::string parts;
    std::mutex mutex_;
};

/**
 * @brief The Builder interface declares methods for creating parts of a complex object.
 */
class Builder {
public:
    virtual ~Builder() {}

    /**
     * @brief Method to build part A.
     */
    virtual void buildPartA() = 0;

    /**
     * @brief Method to build part B.
     */
    virtual void buildPartB() = 0;

    /**
     * @brief Method to retrieve the constructed product.
     * 
     * @return std::shared_ptr<Product> The constructed product.
     */
    virtual std::shared_ptr<Product> getProduct() = 0;
};

/**
 * @brief The ConcreteBuilder class implements the Builder interface to construct a specific product.
 */
class ConcreteBuilder : public Builder {
public:
    /**
     * @brief Constructor to initialize the product.
     */
    ConcreteBuilder() : product(std::make_shared<Product>()) {}

    /**
     * @brief Method to build part A.
     */
    void buildPartA() override {
        product->addPart("Part A");
    }

    /**
     * @brief Method to build part B.
     */
    void buildPartB() override {
        product->addPart("Part B");
    }

    /**
     * @brief Method to retrieve the constructed product.
     * 
     * @return std::shared_ptr<Product> The constructed product.
     */
    std::shared_ptr<Product> getProduct() override {
        return product;
    }

private:
    std::shared_ptr<Product> product;
};

/**
 * @brief The Director class is responsible for using the builder to construct a complex object.
 */
class Director {
public:
    /**
     * @brief Method to set the builder to use.
     * 
     * @param builder The builder to set.
     */
    void setBuilder(std::shared_ptr<Builder> builder) {
        this->builder = builder;
    }

    /**
     * @brief Method to construct the product using the assigned builder.
     */
    void construct() {
        builder->buildPartA();
        builder->buildPartB();
    }

private:
    std::shared_ptr<Builder> builder;
};

int main() {
    std::shared_ptr<Builder> builder = std::make_shared<ConcreteBuilder>();
    Director director;
    director.setBuilder(builder);

    std::vector<std::thread> threads;

    auto threadFunc = [&director, &builder]() {
        director.construct();
        std::shared_ptr<Product> product = builder->getProduct();
        product->showParts();
    };

    for (int i = 0; i < 5; ++i) {
        threads.emplace_back(threadFunc);
    }

    for (auto& thread : threads) {
        thread.join();
    }

    return 0;
}
